crates/polars-time/src/period_start.rs

```diff
@@ -0,0 +1,249 @@
+use arrow::legacy::time_zone::Tz;
+use chrono::{Datelike, NaiveDate, NaiveDateTime, NaiveTime, Timelike};
+use polars_core::prelude::*;
+use polars_core::utils::arrow::temporal_conversions::{
+    timestamp_ms_to_datetime, timestamp_ns_to_datetime, timestamp_us_to_datetime, MILLISECONDS,
+    SECONDS_IN_DAY,
+};
+
+use crate::Duration;
+
+pub(crate) enum RollUnit {
+    Month,
+    Quarter,
+    Year,
+}
+
+#[cfg(feature = "timezones")]
+use crate::utils::{try_localize_datetime, unlocalize_datetime};
+
+// roll backward to the first day of the month
+pub(crate) fn roll_backward(
+    t: i64,
+    tz: Option<&Tz>,
+    timestamp_to_datetime: fn(i64) -> NaiveDateTime,
+    datetime_to_timestamp: fn(NaiveDateTime) -> i64,
+    roll_unit: RollUnit,
+) -> PolarsResult<i64> {
+    let ts = match tz {
+        #[cfg(feature = "timezones")]
+        Some(tz) => unlocalize_datetime(timestamp_to_datetime(t), tz),
+        _ => timestamp_to_datetime(t),
+    };
+    let date = match roll_unit {
+        RollUnit::Month => NaiveDate::from_ymd_opt(ts.year(), ts.month(), 1).ok_or_else(|| {
+            polars_err!(
+                ComputeError: format!("Could not construct date {}-{}-1", ts.year(), ts.month())
+            )
+        })?,
```

Comment:
codecov is showing these errors an uncovered

I don't really how they could error anyway, as we're constructing naive dates where the `date` component is `1`. How about just using `.expect("we know we're creating a valid date")`?